docs/model_classes/usermodel.rst

```diff
@@ -21,17 +21,194 @@ external code.
+    >>> linear.calc([2.], np.array([0, 2, 3]), np.array([2, 3, 5]))
+    array([2., 5., 8.])
+
+In the examples below, we will set up full data classes and fits and not just pass
+the numbers directly into the ``calc`` method.
+
 A one-dimensional model
 =======================
 
-An example is the
+An example is a function similar to the
 `AstroPy trapezoidal model <https://docs.astropy.org/en/stable/api/astropy.modeling.functional_models.Trapezoid1D.html>`_,
 which has four parameters: the amplitude of the central region, the center
 and width of this region, and the slope. The following model class,
-which was not written for efficiancy or robustness, implements this
-interface:
+which was not written for efficiency or robustness, implements this
+interface::
+
+    >>> def _trap1d(pars, x):
+    ...     """Evaluate the Trapezoid.
+    ...
+    ...     Parameters
+    ...     ----------
+    ...     pars: sequence of 4 numbers
+    ...         The order is amplitude, center, width, and slope.
+    ...         These numbers are assumed to be valid (e.g. width
+    ...         is 0 or greater).
+    ...     x: sequence of numbers
+    ...         The grid on which to evaluate the model. It is expected
+    ...         to be a floating-point type.
+    ...
+    ...     Returns
+    ...     -------
+    ...     y: sequence of numbers
+    ...         The model evaluated on the input grid.
+    ...     """
+    ...
+    ...     (amplitude, center, width, slope) = pars
+    ...
+    ...     # There are five segments:
+    ...     #    xlo = center - width/2
+    ...     #    xhi = center + width/2
+    ...     #    x0  = xlo - amplitude/slope
+    ...     #    x1  = xhi + amplitude/slope
+    ...     #
+    ...     #    flat   xlo <= x < xhi
+    ...     #    slope  x0 <= x < xlo
+    ...     #           xhi <= x < x1
+    ...     #    zero   x < x0
+    ...     #           x >= x1
+    ...     #
+    ...     hwidth = width / 2.0
+    ...     dx = amplitude / slope
+    ...     xlo = center - hwidth
+    ...     xhi = center + hwidth
+    ...     x0 = xlo - dx
+    ...     x1 = xhi + dx
+    ...
+    ...     out = np.zeros(x.size)
+    ...     out[(x >= xlo) & (x < xhi)] = amplitude
+    ...
+    ...     idx = np.where((x >= x0) & (x < xlo))
+    ...     out[idx] = slope * x[idx] - slope * x0
+    ...
+    ...     idx = np.where((x >= xhi) & (x < x1))
+    ...     out[idx] = - slope * x[idx] + slope * x1
+    ...
+    ...     return out
+
+    >>> class Trap1D(model.RegriddableModel1D):
+    ...     """A one-dimensional trapezoid.
+    ...
+    ...     The model parameters are:
+    ...
+    ...     ampl
+    ...        The amplitude of the central (flat) segment (zero or greater).
+    ...     center
+    ...        The center of the central segment.
+    ...     width
+    ...        The width of the central segment (zero or greater).
+    ...     slope
+    ...         The gradient of the slopes (zero or greater).
+    ...     """
+    ...
```

Comment:
So, this code/documentation was written before (I hope before) #900 landed, which added some degree of dimension-checking to models. We should probably add a

```python
ndim = 1
```

definition here (and `ndim = 2` to the 2D model). And then mention it earlier in the text.

Note that this is (currently) only used to ensure that models are combined corrcetly (so you can't add 1D + 2D models), and not when it cmoes to model evaluation - that is, to check the arguments that are sent in (because it's hard to do, unless we had something like the structured array approach talked about in #1796 might help).

Reply:
These examples derive from `RegriddableModel1D/2D` and don't have to specify `ndim` because they inherit it form their parent class:
```Python

In [1]: from sherpa.models import model
   ...: class LinearModel(model.RegriddableModel1D):
   ...:     def __init__(self, name='linemodel'):
   ...:         self.slope = model.Parameter(name, 'slope', 1, min=-10, hard_min=-1e20, max=10, hard_max=1e20)
   ...:         super().__init__(name, (self.slope, ))
   ...: 
   ...:     def calc(self, pars, *args, **kwargs):
   ...:         if len(args) == 1:
   ...:             return args[0] * pars[0]
   ...:         elif len(args) == 2:
   ...:             xlo, xhi = args
   ...:             return (xlo + xhi) / 2 * pars[0]
   ...:         else:
   ...:             raise ValueError('This model only works in 1 dimension')
   ...: 

In [2]: import numpy as np
   ...: linear = LinearModel()
   ...: 

In [3]: linear.ndim
Out[3]: 1
```

That's why I did not set it here; however, I'll take the suggestion to add it in the text and explain (a) what it is and (b) why it's not set here.

Reply:
OK, I think it stems from a missing black line higher up (something like "sphinx warning: bulleted list did not end on empty line"), which confused it. It's fixed in my local version, I'll check again after I push.

Reply:
I'd forgotten that I'd added that default to `RegriddableModel1D/2D`.

For background/any-one reading at  home, we use it in the `*OpModel` class to check when creating a combined model - so you'd get an error when you try to combine

```python
linear + sherpa.models.basic.Gauss2D()
```